### PR TITLE
Fix issues with passing byte values as defaultable to a function 

### LINKED ISSUE #29172: Passing a byte value as defaultable to a Ballerina function gives Bad, sad error.
**Description:**
$Subject

**Steps to reproduce:**
```
public function main() {
     testFunc(10);
}

function testFunc(byte b, byte c = 10) {
     byte a = <byte> (b + c);
}
```

Error log:
```
[2021-03-12 14:30:44,437] SEVERE {b7a.log.crash} - null 
java.lang.ArrayIndexOutOfBoundsException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:661)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generateModuleClasses(JvmPackageGen.java:452)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generate(JvmPackageGen.java:438)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:89)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:67)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:267)
	at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:398)
	at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:327)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:145)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:131)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:106)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:129)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:105)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:69)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:251)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
	at org.objectweb.asm.Frame.merge(Frame.java:1268)
	at org.objectweb.asm.Frame.merge(Frame.java:1233)
	at org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1607)
	at org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1543)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.genJMethodForBFunc(MethodGen.java:259)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.generateMethod(MethodGen.java:154)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.lambda$generateModuleClasses$0(JvmPackageGen.java:502)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1746)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```
**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**
Type/Bug, Team/jBallerina

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #29172

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal [hunk 1]
```diff
@@ -190,6 +191,22 @@ function foo7(string s, float b = 1.1, FooObject o = new("default", 100)) return
     return [s, b, o];
 }
 
+function testFuncWithDefaultByteValue() {
+    byte b = funcWithDefaultByteValue();
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal:195]: Shall we add test for call testFuncWithDefaultByteValue with some value and test that value.

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal [hunk 2]
```diff
@@ -190,6 +191,22 @@ function foo7(string s, float b = 1.1, FooObject o = new("default", 100)) return
...
+    test:assertEquals(b, 10);
+}
+
+function testDefaultByteValueInFunctionPointers() {
+    function (byte) returns byte fp = funcWithDefaultByteValue;
+    byte b = fp(15);
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal:201]: Shall we add a test for call testFuncWithDefaultByteValue without value?

**Comment by warunalakshitha** [comment]
Shall we limit the subject line of commit mesaages to 50 characters? 
ex. commit mesaage "Fix issues with passing byte values as defaultable to a function"
Refer https://chris.beams.io/posts/git-commit/#limit-50

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal (hunk 1 above)
Comment (gabilang) [reply tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal]: Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/31612/commits/44689930a2980827b8237aac68183c31e7421e0e

File: tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal (hunk 2 above)
Comment (gabilang) [reply tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal]: Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/31612/commits/9ae22f2369e7867d43fd390bd0fb21a74e7f35a6

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Pull request merged**